### DISCUSSION TITLE
How does $cacheItem->isHit() work?

### DISCUSSION THREAD

**Discussion by lfjeff** [symfony/symfony#47503] [Q&A]
I've been using the Symfony Cache component and seeing some strange behavior with  the`isHit()` method.

For example, I have an application that sets the cache expiration for 300 seconds:
```
$redisConn = RedisAdapter::createConnection('redis://localhost:6379');
$cache = new TagAwareAdapter(new RedisAdapter($redisConn, 'my_cache_name'));
...
$cacheItem = $cache->getItem('my_key');
$cacheItem->set($data);
$cacheItem->expiresAfter(300);
$cache->save($cacheItem);
```

Later, when I call `$cacheItem->isHit()` to check the cache at a time that is well inside the expiration window (say 150 seconds later), it sometimes returns false.  Other times it returns true.  I cannot reliably duplicate the issue.

This code is part of a "cache warmer" script that runs from the command line (normally from `cron`).

I looked at the source code for the method in the `CacheItem` class and all I see is this:
```
public function isHit(): bool
{
    return $this->isHit;
}
```
I don't see how or where `$this->isHit` is ever set.  All I see at the top of the class is `protected $isHit = false`. 

What kind of magic is going on to make this work?  How does `isHit()` determine if the cache is expired?  Why does it sometimes return false when the time is before the cache expire?

**Comment by MatTheCat** [comment]
`$isHit` is set from https://github.com/symfony/symfony/blob/9560d73d8cb815da42bd65e2b7c1c453a2fa4b52/src/Symfony/Component/Cache/Traits/AbstractAdapterTrait.php#L199 and come from the adapterʼs `doFetch` return.

I'm not sure about why `isHit` return seems to be random but it could have to do with the eviction strategy.

Symfony recommends using the `RedisTagAwareAdapter` in your case: https://symfony.com/doc/current/components/cache/adapters/redis_adapter.html#working-with-tags

**Reply by lfjeff** [reply]
Thanks for the reply. Now I know how things work, but unfortunately it did not help.

I've tried using `RedisTagAwareAdapter` in the past, but it causes other issues. I suspected that `TagAwareAdapter` might be causing problems, so I removed it and just used the vanilla `RedisAdapter` (tags are useful, but I don't actually need them for this application).  It also made no difference, I still get random results for `isHit()` when inside the timeout window.

I thought there might be a problem with the `$cache->save($cacheItem)` call, so I put in some debug code to call `isHit()` immediately after the `save()`.  It always returns true.

I'm not sure how to check whether Redis is evicting my keys.  I tried using this command from the Redis CLI:
```
MEMORY USAGE mykey123456
```
and it returns `(nil)` for every key I try.  The Symfony adapter code does a `getId($key)` lookup, so I'm not sure if the key I'm using is the same as the actual Redis key.

My Redis server had a `maxmemory-policy` setting of `noeviction`, so I changed it to `allkeys-lru`, but it makes no difference.

I'm asking Redis to save an array that contains a mix of all kinds of data.  I'm not sure how to determine the size, since I don't know of any PHP function that will return this information (`count()` and `strlen()` are useless for this purpose).  My Redis server does not appear to be starved for memory.  When I run `htop`, it shows only 220M of 4G of RAM being used.

There are no errors in the Redis logs.

Does anyone have any other ideas?
